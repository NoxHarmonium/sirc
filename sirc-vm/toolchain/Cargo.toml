[package]
name = "toolchain"
version = "0.1.0"
edition = "2021"

[dependencies]
clap = { version = "4.3.19", features = ["derive"] }
peripheral_cpu = { path = "../peripheral-cpu" }
nom = "7.1.3"
nom-supreme = "0.8.0"
itertools = "0.11.0"
postcard = { version = "1.0.6", features = ["alloc"] }
serde = "1.0.181"
bytemuck = "1.13.1"
