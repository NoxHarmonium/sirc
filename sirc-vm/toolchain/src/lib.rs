#![warn(clippy::all, clippy::pedantic, clippy::nursery)]
#![allow(
    // I don't like this rule
    clippy::module_name_repetitions,
    // Not sure what this is, will have to revisit
    clippy::must_use_candidate,
    // Will tackle this at the next clean up
    clippy::too_many_lines,
    // Might be good practice but too much work for now
    clippy::missing_errors_doc,
    // Not stable yet - try again later
    clippy::missing_const_for_fn,
    // I have a lot of temporary panics for debugging that will probably be cleaned up
    clippy::missing_panics_doc
)]
#![deny(warnings)]

pub mod data;
pub mod parsers;
pub mod types;
