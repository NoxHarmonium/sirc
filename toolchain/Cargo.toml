[package]
name = "toolchain"
version = "0.1.0"
edition = "2021"

[dependencies]
clap = { version = "3.2.22", features = ["derive"] }
peripheral_cpu = { path = "../peripheral-cpu" }
nom = "7.1.1"
itertools = "0.10.5"
postcard = { version = "1.0.2", features = ["alloc"] }
serde = "1.0.145"

[[bin]]
name = "assembler"
path = "src/assembler.rs"

[[bin]]
name = "linker"
path = "src/linker.rs"
