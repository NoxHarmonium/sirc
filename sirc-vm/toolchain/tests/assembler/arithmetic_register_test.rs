use nom_supreme::{
    error::ErrorTree,
    final_parser::{final_parser, Location},
};
use pretty_hex::{config_hex, HexConfig};
use toolchain::{
    data::object::build_object,
    parsers::instruction::{parse_tokens, Token},
};

#[test]
fn test_assembler_arithmetic_register() {
    let hex_config = HexConfig {
        width: 4,
        ..HexConfig::default()
    };

    let tokens =
        match final_parser::<&str, Vec<Token>, ErrorTree<&str>, ErrorTree<Location>>(parse_tokens)(
            r#"
    ; Two argument - destination and source register the same
    ADDR r1, ph
    ADCR r2, pl
    SUBR r3, sh
    SBCR r4, sl
    ANDR r5, ah
    ORRR r6, al
    XORR r7, lh
    CMPR ll, r7
    TSAR ah, r6
    TSXR al, r5
    COPR sh, r4
    ADDR sl, r3
    ADCR ph, r2
    SUBR pl, r1

    ; Three argument
    ADDR r1, ph, lh
    ADCR r2, pl, ll
    SUBR r3, sh, ah
    SBCR r4, sl, al
    ANDR r5, ah, sh
    ORRR r6, al, sl
    XORR r7, lh, ph
    CMPR ll, r7, r1
    TSAR ah, r6, r2
    TSXR al, r5, r3
    COPR sh, r4, r4
    ADDR sl, r3, r5
    ADCR ph, r2, r6
    SUBR pl, r1, r7
"#,
        ) {
            Ok(tokens) => tokens,
            Err(error) => panic!("Error parsing file:\n{error}"),
        };

    let object = build_object(tokens);
    insta::assert_display_snapshot!(config_hex(&object.program.as_slice(), hex_config));
}
