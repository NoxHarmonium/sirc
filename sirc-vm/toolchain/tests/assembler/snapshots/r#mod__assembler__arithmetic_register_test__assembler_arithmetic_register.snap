---
source: toolchain/tests/assembler/arithmetic_register_test.rs
expression: "config_hex(&object.program.as_slice(), hex_config)"
---
Length: 112 (0x70) bytes
0000:   c0 47 80 10   .G..
0004:   c4 8b c0 10   ....
0008:   c8 cf 00 10   ....
000c:   cd 13 40 10   ..@.
0010:   d1 56 80 10   .V..
0014:   d5 9a c0 10   ....
0018:   d9 de 00 10   ....
001c:   ea 65 c0 10   .e..
0020:   f2 a9 80 10   ....
0024:   fa ed 40 10   ..@.
0028:   ff 31 00 10   .1..
002c:   c3 74 c0 10   .t..
0030:   c7 b8 80 10   ....
0034:   cb fc 40 10   ..@.
0038:   c0 7a 00 10   .z..
003c:   c4 be 40 10   ..@.
0040:   c8 f2 80 10   ....
0044:   cd 36 c0 10   .6..
0048:   d1 6b 00 10   .k..
004c:   d5 af 40 10   ..@.
0050:   d9 e3 80 10   ....
0054:   ea 5c 40 10   .\@.
0058:   f2 98 80 10   ....
005c:   fa d4 c0 10   ....
0060:   ff 11 00 10   ....
0064:   c3 4d 40 10   .M@.
0068:   c7 89 80 10   ....
006c:   cb c5 c0 10   ....
